### PR TITLE
REST API: allow overriding excerpt length

### PR DESCRIPTION
**Pull request by swissspidy**
Addresses https://github.com/WordPress/gutenberg/issues/53570 on the core side

Related:

* https://github.com/WordPress/gutenberg/pull/55400

Trac ticket: https://core.trac.wordpress.org/ticket/59043

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by azaozz** [comment]
The code seems okay but I'm still unsure why this has to work in this particular way. Generally this change makes it so **sometimes** the excerpt length may be overridden with an arbitrary value when doing a REST request. Couple of thoughts about that:

- Seems the initial need was for a hard-coded override of `100` that is only for use in the editor and is later trimmed from JS. Not really sure why this was changed to an arbitrary value as that doesn't seem needed?
- The change seems implemented in a somewhat strange way by adding a non-removable filter callback. As it is still a filter, the excerpt length value can be overridden by adding another filter callback that runs after this one. However is seems that the block editor is (somewhat) hard-coded to expect the exact length of 100 words for the excerpt in edit context, so allowing that to be overridden doesn't seem like a particularly good idea?

How about instead of adding that non-removable filter callback we add another param like `$override_length = null` to `wp_trim_excerpt()` and pass that through the `'get_the_excerpt'` filter that's used right underneath? So 
```
$excerpt = apply_filters( 'get_the_excerpt', $post->post_excerpt, $post );
```
would become:
```
$excerpt = apply_filters( 'get_the_excerpt', $post->post_excerpt, $post, (int) $override_excerpt_length );
```
where `$override_excerpt_length` may actually be better off as a hard-coded `100`. Untested, but thinking this would work perhaps a bit better? At least will enforce the length when set through the REST API and perhaps prevent "strange" accidental edge cases in the future if the `excerpt_length` filter is used by a plugin that "must enforce its own setting at all costs" :)

**Comment by swissspidy** [comment]
> However is seems that the block editor is (somewhat) hard-coded to expect the exact length of 100 words for the excerpt in edit context [...]

How did you get to this conclusion, where did you get this number from? 🤔 

The whole issue right now is that Gutenberg offers a way to fully customize the excerpt length in the editor, but that length is not actually honored in the REST API response it receives.
Screenshot 2024-01-23 at 23 49 21

This PR is an attempt to solve this in a less hacky way than https://github.com/WordPress/gutenberg/pull/55400

> As it is still a filter, the excerpt length value can be overridden by adding another filter callback that runs after this one.

Theoretically, yes. With `PHP_INT_MAX` this risk is greatly reduced though.

But plugins can also completely override the excerpt using the `get_the_excerpt` filter anyway. There's no safeguard for that.

> How about instead of adding that non-removable filter callback we add another param like $override_length = null to `wp_trim_excerpt()` and pass that through the 'get_the_excerpt' filter that's used right underneath? So
>  At least will enforce the length when set through the REST API and perhaps prevent "strange" accidental edge cases in the future if the excerpt_length filter is used by a plugin that "must enforce its own setting at all costs" :)

This approach does not prevent a plugin from filtering `get_the_excerpt` and enforcing its own setting at all costs.

So in this regard, both approaches have the same problem and I don't think it's solvable. Plugins will always be able to filter the excerpt.

**Comment by azaozz** [comment]
> How did you get to this conclusion, where did you get this number from? 🤔

The initial bug report, [here](https://github.com/WordPress/gutenberg/issues/53570), and the initial patch? :) 

> The whole issue right now is that Gutenberg offers a way to fully customize the excerpt length in the editor, but that length is not actually honored in the REST API response it receives.

May be missing something but how would that slider work if the excerpt was too short? Lets say the user sets the excerpt length to 20, how can they increase it next time they edit the post? Afaik this is the problem here, the editor needs a consistent excerpt length (100?) that can be trimmed on display to allow the user to change it (and actually see it).

> This approach does not prevent a plugin from filtering get_the_excerpt

Right. But that most likely would be a deliberate change targeting exactly this functionality. I'm thinking more about accidental overrides by plugins that do filters with `PHP_INT_MAX`, etc.

**Comment by swissspidy** [comment]
> The initial bug report, https://github.com/WordPress/gutenberg/issues/53570, and the initial patch? :)

There was no _need_ for a _hardcoded_ excerpt length. That was simply a hacky workaround to provide something to the editor to work with. 100 is the highest number that can be chosen in the editor. No proper solution was implemented which actually returns the actually configured length.
So I am offering this PR to fix that, as it allows trimming excerpts at the desired maximum length.

> May be missing something but how would that slider work if the excerpt was too short?

Just like the `excerpt_length` filter, the whole setting in the editor is a _maximum_ excerpt length. If your `post_content` is shorter, you'll get a shorter excerpt.

>  Lets say the user sets the excerpt length to 20, how can they increase it next time they edit the post? 

The excerpt block is used in places like Query Loop blocks. If you want to increase the excerpt length in places where you used the block, go edit the block.

But questions about the excerpt block are better target at someone who built the block :)
I'm merely offering help here to fix a bad implementation.

> Afaik this is the problem here, the editor needs a consistent excerpt length (100?) that can be trimmed on display to allow the user to change it (and actually see it).

What the editor needs is a way to trim the excerpt length to the desired maximum and this change here allows it to do so. At https://github.com/WordPress/gutenberg/pull/55400#issuecomment-1905838055 I explained why it's better to do the trimming server-side rather than client-side.

**Comment by draganescu** [comment]
We also need a Gutenberg PR to use this new param. @azaozz I think this approach is the best since it solves the _actual_ problem: one block wants excerpt to be size X for itself when it renders and when it displays. It should not do this in any way that has any potential to leak the setting to other operations involving excerpt.

Having a special param gives the block this option. I also think that the fact that in the end WP can be configured in ways which make this param not work as expected is something to be dealt with at the end of the line - at that implementation: e.g. if your excerpt modifying plugin breaks how you work with the block deal with the plugin.

Plugins themselves can also check the request and not override it so it's easy to adapt.

**Comment by zobinadevain** [comment]
This approach does not prevent a plugin from filtering get_the_excerpt and enforcing its own setting at all costs.
So in this regard, both approaches have the same problem and I don't think it's solvable. Plugins will always be able to filter the excerpt.

**Comment by zobinadevain** [comment]
<a href='https://capmistoi-6241.weebly.com'>.</a>
<a href='https://capmistoi-6242.weebly.com'>.</a>
<a href='https://capmistoi-6243.weebly.com'>.</a>
<a href='https://capmistoi-6244.weebly.com'>.</a>
<a href='https://capmistoi-6245.weebly.com'>.</a>
<a href='https://capmistoi-6246.weebly.com'>.</a>
<a href='https://capmistoi-6247.weebly.com'>.</a>
<a href='https://capmistoi-6248.weebly.com'>.</a>
<a href='https://capmistoi-6249.weebly.com'>.</a>
<a href='https://capmistoi-6250.weebly.com'>.</a>
<a href='https://capmistoi-6251.weebly.com'>.</a>
<a href='https://capmistoi-6252.weebly.com'>.</a>
<a href='https://capmistoi-6253.weebly.com'>.</a>
<a href='https://capmistoi-6254.weebly.com'>.</a>
<a href='https://capmistoi-6255.weebly.com'>.</a>
<a href='https://capmistoi-6256.weebly.com'>.</a>
<a href='https://capmistoi-6257.weebly.com'>.</a>
<a href='https://capmistoi-6258.weebly.com'>.</a>
<a href='https://capmistoi-6259.weebly.com'>.</a>
<a href='https://capmistoi-6260.weebly.com'>.</a>
<a href='https://capmistoi-6261.weebly.com'>.</a>
<a href='https://capmistoi-6262.weebly.com'>.</a>
<a href='https://capmistoi-6263.weebly.com'>.</a>
<a href='https://capmistoi-6264.weebly.com'>.</a>
<a href='https://capmistoi-6265.weebly.com'>.</a>
<a href='https://capmistoi-6266.weebly.com'>.</a>
<a href='https://capmistoi-6267.weebly.com'>.</a>
<a href='https://capmistoi-6268.weebly.com'>.</a>
<a href='https://capmistoi-6269.weebly.com'>.</a>
<a href='https://capmistoi-6270.weebly.com'>.</a>

File: tests/phpunit/tests/rest-api/rest-posts-controller.php [hunk 1]
```diff
@@ -2345,6 +2344,33 @@ public function test_prepare_item_skips_content_filter_if_not_needed() {
...
+		);
+
+		$endpoint = new WP_REST_Posts_Controller( 'post' );
+		$request  = new WP_REST_Request( 'GET', sprintf( '/wp/v2/posts/%d', $post_id ) );
+		$request->set_param( 'context', 'edit' );
+		$request->set_param( '_fields', 'excerpt' );
+		$request->set_param( 'excerpt_length', 43 );
+		$response = $endpoint->prepare_item_for_response( get_post( $post_id ), $request );
+		$data     = $response->get_data();
+		$this->assertArrayHasKey( 'excerpt', $data );
```
Comment (anton-vlasenko) [inline tests/phpunit/tests/rest-api/rest-posts-controller.php]:
```suggestion
		$this->assertArrayHasKey( 'excerpt', $data, 'Response must contain an "excerpt" key.' );
```

File: tests/phpunit/tests/rest-api/rest-posts-controller.php [hunk 2]
```diff
@@ -2345,6 +2344,33 @@ public function test_prepare_item_skips_content_filter_if_not_needed() {
...
+
+		// 43 words plus the ellipsis added via the 'excerpt_more' filter.
+		$this->assertCount( 44, explode( ' ', $data['excerpt']['rendered'] ) );
```
Comment (anton-vlasenko) [inline tests/phpunit/tests/rest-api/rest-posts-controller.php]:
```suggestion
		$this->assertCount( 44,
			explode( ' ', $data['excerpt']['rendered'] ),
			'Incorrect word count in the excerpt. Expected the excerpt to contain 44 words (43 words plus an ellipsis), but a different word count was found.'
		);
```

File: tests/phpunit/tests/rest-api/rest-posts-controller.php [hunk 3]
```diff
@@ -2345,6 +2344,33 @@ public function test_prepare_item_skips_content_filter_if_not_needed() {
 		$this->assertSame( 0, $filter_count );
 	}
 
+	public function test_prepare_item_override_excerpt_length() {
```
Comment (anton-vlasenko) [inline tests/phpunit/tests/rest-api/rest-posts-controller.php:2370]:
```suggestion
	/**
	 * @ticket 59043
	 *
	 * @covers WP_REST_Posts_Controller::prepare_item_for_response
	 */
	public function test_prepare_item_override_excerpt_length() {
```

**Review by anton-vlasenko — COMMENTED** [review]
Overall, this looks good to me.
I have some minor suggestions.

File: tests/phpunit/tests/rest-api/rest-posts-controller.php (hunk 1 above)
Comment (swissspidy) [reply tests/phpunit/tests/rest-api/rest-posts-controller.php]: tbh I could probably just remove this assertion here, it's a bit superfluous

**Comment by moshiajonta** [comment]
You are asking right but you can see few points on there i collected them from repurchase
https://freelancewebdesignerdubai.com/
https://reloaderschoiceusa.com/
https://viralnpc.com/
http://humanedivorce.com/

**Comment by moshiajonta** [comment]
First i was thing same like you but right-now there are few things maybe you will like it
<a href='https://capmistoi-6221.weebly.com'>capmistoi-6221.weebly.com
</a>
<a href='https://capmistoi-6222.weebly.com'>capmistoi-6222.weebly.com
</a>
<a href='https://capmistoi-6223.weebly.com'>capmistoi-6223.weebly.com
</a>
<a href='https://capmistoi-6224.weebly.com'>capmistoi-6224.weebly.com
</a>
<a href='https://capmistoi-6225.weebly.com'>capmistoi-6225.weebly.com
</a>
<a href='https://capmistoi-6226.weebly.com'>capmistoi-6226.weebly.com
</a>
<a href='https://capmistoi-6227.weebly.com'>capmistoi-6227.weebly.com
</a>
<a href='https://capmistoi-6228.weebly.com'>capmistoi-6228.weebly.com
</a>
<a href='https://capmistoi-6229.weebly.com'>capmistoi-6229.weebly.com
</a>
<a href='https://capmistoi-6230.weebly.com'>capmistoi-6230.weebly.com
</a>
<a href='https://capmistoi-6231.weebly.com'>capmistoi-6231.weebly.com
</a>
<a href='https://capmistoi-6232.weebly.com'>capmistoi-6232.weebly.com
</a>
<a href='https://capmistoi-6233.weebly.com'>capmistoi-6233.weebly.com
</a>
<a href='https://capmistoi-6234.weebly.com'>capmistoi-6234.weebly.com
</a>
<a href='https://capmistoi-6235.weebly.com'>capmistoi-6235.weebly.com
</a>
<a href='https://capmistoi-6236.weebly.com'>capmistoi-6236.weebly.com
</a>
<a href='https://capmistoi-6237.weebly.com'>capmistoi-6237.weebly.com
</a>
<a href='https://capmistoi-6238.weebly.com'>capmistoi-6238.weebly.com
</a>
<a href='https://capmistoi-6239.weebly.com'>capmistoi-6239.weebly.com
</a>
<a href='https://capmistoi-6240.weebly.com'>capmistoi-6240.weebly.com
</a>

File: tests/phpunit/tests/rest-api/wpRestMenuItemsController.php [hunk 4]
```diff
@@ -156,8 +156,8 @@ public function test_registered_get_item_params() {
 		$response = rest_get_server()->dispatch( $request );
 		$data     = $response->get_data();
 		$keys     = array_keys( $data['endpoints'][0]['args'] );
-		sort( $keys );
-		$this->assertSame( array( 'context', 'id' ), $keys );
+		$keys     = array_keys( $data['endpoints'][0]['args'] );
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/rest-api/wpRestMenuItemsController.php]:
```suggestion
```

Duplicate code. Same code present in above line.

**Comment by swissspidy** [comment]
@azaozz I would appreciate another review

If we can merge this now into core there is still time to get the corresponding change into this week's Gutenberg RC so we can finally  fix this in 6.5

**Comment by TimothyBJacobs** [comment]
This is a solution that I don't love, but can't think of anything better.

The reason I don't love it, is it feels different to how we handle everything else that is variable (like image sizes, avatar sizes, raw vs rendered) where the API spits out multiple choices, and the client chooses the one that is closest to their desired size. I think this is the first time we're giving the client control over a specific API property like that.

But that all being said, I can't really think of a better way of handling this. And I think this can be thought of as an extension to things like `?_fields`.

**Comment by draganescu** [comment]
With the latest thumbs up this is definitely ready IMO :)

**Comment by swissspidy** [comment]
Committed in https://core.trac.wordpress.org/changeset/58065

**Pull request closed without merging**